Refactor: split unit tests by API resources

nejch opened the pull request:
I've been thinking about how to make testing more accessible for contributors and maybe splitting them by resources in the Gitlab API docs would be best because that's how most people interact with this anyway. If there's a new API resource it likely gets a new page so this would kind of make it a 1-to-1 for new contributions.

Ideas:
- [x] Split unit tests by API resources
- [x] Convert unittest classes into modules? (disclaimer: really biased here against classes :P) 
- [x] ~~Figure out a better way for fixtures & mocking copy/pasting~~

max-wittig commented on gitlab/__init__.py:
```diff
@@ -321,7 +311,7 @@ def set_license(self, license, **kwargs):
         if "next_url" in parameters:
             return parameters["next_url"]
-        args = _sanitize(parameters)
+        args = utils.sanitize_parameters(parameters)
```
nice change. Get rid of custom functions!

nejch commented on gitlab/tests/objects/test_application.py:
```diff
@@ -1,120 +1,131 @@
+        path="/api/v4/application/appearance",
+        method="put",
+    )
+    def resp_update_appearance(url, request):
```
Oops, probably some leftovers from the copy/paste. Late-night commits :grin: I've taken a few more duplicate responses out of other tests now so they can be shared where possible.

nejch commented on gitlab/tests/objects/test_users.py:
```diff
@@ -0,0 +1,102 @@
+def resp_get_user_memberships(url, request):
+    headers = {"content-type": "application/json"}
+    content = """[
+                  {
```
Thanks! I've made it look more like the mirrors example response now.

nejch commented:
> Looks good overall so far!
> 
> I'm also biased against classes in python. See also the last sentence of the Zen of Python

:grin: that's great I'm glad, I was afraid these changes would be too much. I've only left the 2 classes in `test_base.py` for now as it's a small file and didn't seem to make sense to split it yet.

I realized I might be getting myself into a continuous rebase party as new PR's with tests in the old files are merged, so I've cleaned it up a bit more and I think I want to look at fixtures in a later PR when I have more time, if this is in an OK shape?

max-wittig commented:
@nejch Sorry totally missed this. Does one only see assignments and not reviewers in the feed?

max-wittig commented on gitlab/tests/mixins/test_object_mixins_attributes.py:
```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
```
Where does this license header come from?

nejch replied:
Oops I must have left this when splitting the files from the original mixin tests. I've removed it now but I can also add new headers to all the new files if needed.

max-wittig commented:
@nejch I would love to merge this. Could you please rebase to latest master?

nejch commented:
> 
> 
> @nejch I would love to merge this. Could you please rebase to latest master?

Sorry @max-wittig I'm back from the dead/vacation/other distractions now 😁 I've rebased and rewritten the tests using responses now so it's more consistent with your runners example. Could you have another look to see if this is how you imagined `responses`-based tests?

Not sure about notifications, there's a separate Notification setting for PR Reviews so maybe you had them turned off?

max-wittig commented:
We actually have less tests with this MR. I guess some got lost. Could yout take a look? @nejch 

This MR: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720445167
Master: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720460657

nejch commented:
> We actually have less tests with this MR. I guess some got lost. Could yout take a look? @nejch
> 
> 155 vs. 158
> 
> This MR: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720445167
> Master: https://travis-ci.org/github/python-gitlab/python-gitlab/jobs/720460657

Oof, I was weirded out by this until I realized that on master, 2 test cases (in `test_groups.py`) were run 3 times. They're duplicates because I had accidentally put them in a unittest base class (TestGroup) inherited by other test classes back when I was adding group export/import tests. :facepalm: 
With pytest they're run only once, hence the lower number. If you exclude `test_groups.py` the number should be higher. But I checked all cases collected in master vs. this PR again and it should be OK otherwise.

With `tox -e py38 -- -v`:

master:
```
gitlab/tests/objects/test_groups.py::TestGroup::test_create_group PASSED [ 75%]
gitlab/tests/objects/test_groups.py::TestGroup::test_get_group PASSED    [ 75%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_create_group PASSED [ 76%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_create_group_export PASSED [ 76%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_download_group_export PASSED [ 77%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_get_group PASSED [ 78%]
gitlab/tests/objects/test_groups.py::TestGroupExport::test_refresh_group_export_status SKIPPED [ 78%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_create_group PASSED [ 79%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_get_group PASSED [ 80%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_import_group PASSED [ 80%]
gitlab/tests/objects/test_groups.py::TestGroupImport::test_refresh_group_import_status SKIPPED [ 81%]
```

refactor:
```
gitlab/tests/objects/test_groups.py::test_get_group PASSED               [ 56%]
gitlab/tests/objects/test_groups.py::test_create_group PASSED            [ 56%]
gitlab/tests/objects/test_groups.py::test_create_group_export PASSED     [ 57%]
gitlab/tests/objects/test_groups.py::test_refresh_group_export_status SKIPPED [ 57%]
gitlab/tests/objects/test_groups.py::test_download_group_export PASSED   [ 58%]
gitlab/tests/objects/test_groups.py::test_import_group PASSED            [ 58%]
gitlab/tests/objects/test_groups.py::test_refresh_group_import_status SKIPPED [ 59%]
```

max-wittig commented:
Thank you very much! Very nice!
